src/mempool_limit.cpp

```diff
@@ -0,0 +1,160 @@
+    int rightWeight = getWeightAt(fromIndex).evictionWeight - getWeightAt(fromIndex * 2 + 2).evictionWeight;
+    // On Left
+    if (weightToFind < leftWeight) {
+        return findByEvictionWeight(fromIndex * 2 + 1, weightToFind);
+    }
+    // Found
+    if (weightToFind < rightWeight) {
+        return fromIndex;
+    }
+    // On Right
+    return findByEvictionWeight(fromIndex * 2 + 2, weightToFind - rightWeight);
+}
+
+TxWeight WeightedTxTree::getTotalWeight() const
+{
+    return getWeightAt(0);
+}
+
+
+void WeightedTxTree::add(const WeightedTxInfo& weightedTxInfo)
+{
+    if (txIdToIndexMap.count(weightedTxInfo.txId) > 0) {
+        // This should not happen, but should be prevented nonetheless
+        return;
+    }
+    txIdAndWeights.push_back(weightedTxInfo);
+    childWeights.push_back(ZERO_WEIGHT);
+    txIdToIndexMap[weightedTxInfo.txId] = size;
+    backPropagate(size, weightedTxInfo.txWeight);
+    size += 1;
+}
+
+void WeightedTxTree::remove(const uint256& txId)
+{
+    if (txIdToIndexMap.find(txId) == txIdToIndexMap.end()) {
+        // Remove may be called multiple times for a given tx, so this is necessary
+        return;
+    }
+
+    size_t removeIndex = txIdToIndexMap[txId];
+
+    // We reduce the size at the start of this method to avoid saying size - 1
+    // when refering to the last element of the array below
+    size -= 1;
+
+    TxWeight lastChildWeight = txIdAndWeights[size].txWeight;
+    backPropagate(size, lastChildWeight.negate());
+
+    if (removeIndex < size) {
+        TxWeight weightDelta = lastChildWeight.add(txIdAndWeights[removeIndex].txWeight.negate());
+        txIdAndWeights[removeIndex] = txIdAndWeights[size];
+        txIdToIndexMap[txIdAndWeights[removeIndex].txId] = removeIndex;
+        backPropagate(removeIndex, weightDelta);
+    }
+
+    txIdToIndexMap.erase(txId);
+    txIdAndWeights.pop_back();
+    childWeights.pop_back();
+}
+
+boost::optional<uint256> WeightedTxTree::maybeDropRandom()
+{
+    TxWeight totalTxWeight = getTotalWeight();
+    if (totalTxWeight.cost <= capacity) {
+        return boost::none;
+    }
+    LogPrint("mempool", "Mempool cost limit exceeded (cost=%d, limit=%d)\n", totalTxWeight.cost, capacity);
+    int randomWeight = GetRand(totalTxWeight.evictionWeight);
+    WeightedTxInfo drop = txIdAndWeights[findByEvictionWeight(0, randomWeight)];
+    LogPrint("mempool", "Evicting transaction (txid=%s, cost=%d, evictionWeight=%d)\n",
+        drop.txId.ToString(), drop.txWeight.cost, drop.txWeight.evictionWeight);
+    remove(drop.txId);
+    return drop.txId;
+}
+
+
+TxWeight TxWeight::add(const TxWeight& other) const
+{
+    return TxWeight(cost + other.cost, evictionWeight + other.evictionWeight);
+}
+
+TxWeight TxWeight::negate() const
+{
+    return TxWeight(-cost, -evictionWeight);
+}
+
+
+WeightedTxInfo WeightedTxInfo::from(const CTransaction& tx, const CAmount& fee)
+{
+    size_t memUsage = RecursiveDynamicUsage(tx);
```

Comment:
Shouldn't `RecursiveDynamicUsage` be changed to include the sizes of JoinSplit, Output, and Spend descriptions?

Reply:
Yes, I will file a separate ticket for this.

Reply:
https://github.com/zcash/zcash/issues/4158